sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/InputsAdapter.java

```diff
@@ -250,4 +240,95 @@ private boolean isMatchingExtRef(TCompasFlow compasFlow, TExtRef extRef) {
     private LDeviceAdapter getLDeviceAdapter() {
         return parentAdapter.getParentAdapter();
     }
+
+    public List<SclReportItem> updateAllSourceDataSetsAndControlBlocks() {
+        List<TCompasFlow> compasFlows = PrivateService.extractCompasPrivates(currentElem, TCompasFlow.class);
+        String currentBayUuid = getIedAdapter().getPrivateCompasBay().map(TCompasBay::getUUID).orElse(null);
+        if (StringUtils.isBlank(currentBayUuid)) {
+            return List.of(getIedAdapter().buildFatalReportItem(MESSAGE_IED_IS_MISSING_PRIVATE_COMPAS_BAY_UUID));
+        }
+        return getExtRefs().stream()
+            .filter(this::areBindingAttributesPresent)
+            .filter(this::isExternalBound)
+            //TODO: replace filter below by criteria of issue #93 RSR-389
+            // - Active or Untested i.e /IED/LDevice/LN0/Inputs/Private/compas:Flow @FlowStatus="ACTIVE" OR "UNTESTED"
+            .filter(extRef -> !getMatchingCompasFlows(extRef, compasFlows).isEmpty())
+            .map(extRef -> updateSourceDataSetsAndControlBlocks(extRef, currentBayUuid))
+            .flatMap(Optional::stream)
+            .collect(Collectors.toList());
+    }
+
+    private boolean isExternalBound(TExtRef tExtRef) {
+        return !tExtRef.getIedName().equals(getIedAdapter().getName());
+    }
+
+    private boolean areBindingAttributesPresent(TExtRef tExtRef) {
+        return StringUtils.isNotBlank(tExtRef.getIedName())
+            && StringUtils.isNotBlank(tExtRef.getDesc())
+            && StringUtils.isNotBlank(tExtRef.getLdInst())
+            && tExtRef.getLnClass().stream().findFirst().filter(StringUtils::isNotBlank).isPresent()
+            && StringUtils.isNotBlank(tExtRef.getDoName());
+    }
+
+    private Optional<SclReportItem> updateSourceDataSetsAndControlBlocks(TExtRef extRef, String targetBayUuid) {
+        if (extRef.getServiceType() == null) {
+            return fatalReportItem(extRef, MESSAGE_SERVICE_TYPE_MISSING_ON_EXTREF);
+        }
+        Optional<IEDAdapter> optionalSrcIedAdapter = getSclRootAdapter().findIedAdapterByName(extRef.getIedName());
+        if (optionalSrcIedAdapter.isEmpty()) {
+            return fatalReportItem(extRef, "Source IED not found in SCD");
+        }
+        IEDAdapter sourceIed = optionalSrcIedAdapter.get();
+        Optional<String> sourceIedBayUuid = sourceIed.getPrivateCompasBay().map(TCompasBay::getUUID);
+        if (sourceIedBayUuid.isEmpty() || StringUtils.isBlank(sourceIedBayUuid.get())) {
```

Comment:
Why this 2 checks

Reply:
Actually, there is a better way to this with `Optional.filter`. I suggest : 
```java
        Optional<String> sourceIedBayUuid = sourceIed.getPrivateCompasBay()
            .map(TCompasBay::getUUID)
            .filter(StringUtils::isNotBlank);
        if (sourceIedBayUuid.isEmpty()) {
            return fatalReportItem(extRef, "Source IED is missing Private/compas:Bay@UUID attribute");
        }
```